src/System.Private.ServiceModel/src/System/ServiceModel/Channels/ProducerConsumerStream.cs

```diff
@@ -0,0 +1,350 @@
+                _readOnly = isReadOnly;
+            }
+
+            public byte[] ByteBuffer
+            {
+                get { return _buffer; }
+                set
+                {
+                    if (_buffer == value)
+                    {
+                        return;
+                    }
+                    if (_readOnly)
+                    {
+                        throw new ArgumentException("value");
+                    }
+                    _buffer = value;
+                }
+            }
+
+            public int Offset
+            {
+                get { return _offset; }
+                set
+                {
+                    if (_offset == value)
+                    {
+                        return;
+                    }
+                    if (_readOnly)
+                    {
+                        throw new ArgumentException("value");
+                    }
+                    _offset = value;
+                }
+            }
+
+            public int Count
+            {
+                get { return _count; }
+                set
+                {
+                    if (_count == value)
+                    {
+                        return;
+                    }
+                    if (_readOnly)
+                    {
+                        throw new ArgumentException("value");
+                    }
+                    _count = value;
+                }
+            }
+
+            public override bool Equals(object obj)
+            {
+                if (obj is WriteBufferWrapper)
+                    return this.Equals((WriteBufferWrapper)obj);
+                return false;
+            }
+
+            public bool Equals(WriteBufferWrapper other)
+            {
+                // Doing _readOnly comparison first as main use case is comparing against
+                // BufferContainer.EmptyContainer which is the only readonly instance of BufferContainer.
+                // If either instance is the read only, then this field is sufficient for comparison.
+                if (_readOnly || other._readOnly)
+                {
+                    return _readOnly == other._readOnly;
+                }
+
+                if (other._buffer == _buffer && other._offset == _offset)
+                {
+                    return other._count == _count;
+                }
+
+                return false;
+            }
+
+            public static bool operator ==(WriteBufferWrapper a, WriteBufferWrapper b)
+            {
+                return a.Equals(b);
+            }
+
+            public static bool operator !=(WriteBufferWrapper a, WriteBufferWrapper b)
+            {
+                return !(a == b);
+            }
+
+            internal int Write(byte[] srcBuffer, int srcOffset, int srcCount)
+            {
+                if (_readOnly)
+                {
+                    throw new InvalidOperationException("BufferContainer.Write");
+                }
```

Comment:
It isn't a public type, but is it reachable from public API's? (Like someone calls some HttpContent public API that ultimately ends up calling this method).  If the answer is 'no', I'd favor Contract.Assert.  BTW, we have an ObjectIsReadOnly resource string if you decide the exception should remain.